pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -58,161 +57,246 @@ type rangefeedMuxer struct {
+		transport, err := newTransportForRange(ctx, token.Desc(), m.ds)
+		if err != nil {
+			log.VErrEventf(ctx, 1, "Failed to create transport for %s ", token.String())
+			continue
+		}
+		releaseTransport = transport.Release
 
-		m.mu.Lock()
-		defer m.mu.Unlock()
+		for !transport.IsExhausted() {
+			args := makeRangeFeedRequest(span, token.Desc().RangeID, m.cfg.overSystemTable, startAfter, m.cfg.withDiff)
+			args.Replica = transport.NextReplica()
+			args.StreamID = streamID
 
-		ms.numStreams--
-		if ms.numStreams == 0 {
-			delete(m.mu.clients, req.Replica.NodeID)
-			if log.V(1) {
-				log.InfofDepth(streamCtx, 1, "shut down inactive mux for node %d", req.Replica.NodeID)
+			rpcClient, err := transport.NextInternalClient(ctx)
+			if err != nil {
+				log.VErrEventf(ctx, 1, "RPC error connecting to replica %s: %s", args.Replica, err)
+				continue
 			}
-			ms.cancel()
+
+			conn, err := m.establishMuxConnection(ctx, rpcClient, args.Replica.NodeID)
+			if err != nil {
+				return err
+			}
+
+			if err := conn.startRangeFeed(streamID, stream, &args); err != nil {
+				log.VErrEventf(ctx, 1,
+					"RPC error establishing mux rangefeed to replica %s: %s", args.Replica, err)
+				continue
+			}
+			// We successfully established rangefeed, so the responsibility
+			// for releasing the stream is transferred to the mux go routine.
+			stream = nil
+			return nil
 		}
-	}
 
-	if err := ms.client.Send(req); err != nil {
-		cleanup()
-		return nil, nil, err
+		// If the transport is exhausted, we evict routing token and retry range
+		// resolution.
+		token.Evict(ctx)
+		token = rangecache.EvictionToken{}
 	}
-	return producer, cleanup, nil
+
+	return ctx.Err()
 }
 
 // establishMuxConnection establishes MuxRangeFeed RPC with the node.
 func (m *rangefeedMuxer) establishMuxConnection(
 	ctx context.Context, client rpc.RestrictedInternalClient, nodeID roachpb.NodeID,
-) (*muxClientState, error) {
-	// NB: the `ctx` in scope here belongs to a client for a single range feed, and must
-	// not influence the lifetime of the mux connection. At the time of writing, the caller
-	// is `singleRangeFeed` which calls into this method through its streamProducerFactory
-	// argument.
-	m.mu.Lock()
-	ms, found := m.mu.clients[nodeID]
-	if !found {
-		// Initialize muxClientState.
-		// Only initCtx is initialized here since we need to block on it.
-		// The rest of the initialization happens in startNodeMuxRangeFeed.
-		ms = &muxClientState{initCtx: makeTerminationContext()}
-		// Kick off client initialization on another Go routine.
-		// It is important that we start MuxRangeFeed RPC using long-lived
-		// context available in the main context group used for this muxer.
-		m.g.GoCtx(func(ctx context.Context) error {
-			return m.startNodeMuxRangeFeed(ctx, ms, client, nodeID)
-		})
-		m.mu.clients[nodeID] = ms
+) (*muxNodeConn, error) {
+	var muxClient *future.Future[connOrError]
+
+	{
+		ptr, exists := m.muxClients.Load(int64(nodeID))
+		if !exists {
```

Comment:
Why don't we call `LoadOrStore` immediately here? Is the future construction expensive enough to matter?